implementations/rust/ockam/ockam_identity/tests/message_flow_auth.rs

```diff
@@ -0,0 +1,316 @@
+        session,
+    };
+
+    Ok(info)
+}
+
+struct TcpConnectionInfo {
+    address: Address,
+    session: Option<(Sessions, SessionId)>,
+}
+
+async fn create_connection(
+    ctx: &Context,
+    socket_addr: &SocketAddr,
+    with_session: bool,
+) -> Result<TcpConnectionInfo> {
+    let tcp = TcpTransport::create(ctx).await?;
+    let (address, session) = if with_session {
+        let sessions = Sessions::default();
+        let session_id = sessions.generate_session_id();
+        let trust_options = TcpConnectionTrustOptions::new().with_session(&sessions, &session_id);
+        let address = tcp.connect(socket_addr.to_string(), trust_options).await?;
+        (address, Some((sessions, session_id)))
+    } else {
+        let address = tcp
+            .connect(socket_addr.to_string(), TcpConnectionTrustOptions::new())
+            .await?;
+        (address, None)
+    };
+
+    let info = TcpConnectionInfo { address, session };
+
+    Ok(info)
+}
+
+struct SecureChannelListenerInfo {
+    identity: Identity<Vault, InMemoryStorage>,
+}
+
+impl SecureChannelListenerInfo {
+    fn get_channel(&self) -> Address {
+        self.identity
+            .secure_channel_registry()
+            .get_channel_list()
+            .first()
+            .unwrap()
+            .encryptor_messaging_address()
+            .clone()
+    }
+}
+
+async fn create_secure_channel_listener(
+    ctx: &Context,
+    session: &Option<(Sessions, SessionId)>,
+) -> Result<SecureChannelListenerInfo> {
+    let identity = Identity::create(ctx, &Vault::create()).await?;
+
+    let trust_options =
+        SecureChannelListenerTrustOptions::new().with_trust_policy(TrustEveryonePolicy);
+    let trust_options = if let Some((sessions, session_id)) = session {
+        trust_options.with_session(sessions, session_id)
+    } else {
+        trust_options
+    };
+
+    identity
+        .create_secure_channel_listener("listener", trust_options)
+        .await?;
+
+    let info = SecureChannelListenerInfo { identity };
+
+    Ok(info)
+}
+
+struct SecureChannelInfo {
+    identity: Identity<Vault, InMemoryStorage>,
+    address: Address,
+}
+
+async fn create_secure_channel(
+    ctx: &Context,
+    connection: &TcpConnectionInfo,
+) -> Result<SecureChannelInfo> {
+    let identity = Identity::create(ctx, &Vault::create()).await?;
+
+    let trust_options = SecureChannelTrustOptions::new();
+    let trust_options = if let Some((sessions, session_id)) = &connection.session {
+        trust_options.with_ciphertext_session(sessions, session_id)
+    } else {
+        trust_options
+    };
+
+    let address = identity
+        .create_secure_channel(
+            route![connection.address.clone(), "listener"],
+            trust_options,
+        )
+        .await?;
+
+    let info = SecureChannelInfo { identity, address };
+
+    Ok(info)
+}
+
+#[allow(non_snake_case)]
+#[ockam_macros::test]
+async fn sessions__secure_channel_over_tcp_without_session__should_pass_messages(
```

Comment:
That's a stylistic thing but personally I'd rather read about the tests first in a test file, then have a look at the implementation, instead of having to read a bunch of implementation-related code until I get to the context where it is used

Reply:
My C/C++ mind sometimes protests to declare functions after their usage. But I agree